Juan Esteban Camacho Castro, Juan Diego Serpa Ruiz

In [1]:
!pip install xlrd

# Semana 9, Codificación de variables categóricas

### Importación de liberías y datos únicos

In [2]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder

In [3]:
df = pd.read_csv("S08_PD_Camacho_Serpa_DatasetEscalado.csv")

In [4]:
df["categoria_producto"].unique()

array(['Moda', 'Electrónica', 'Hogar', 'Belleza', 'Juguetería',
       'Deportes'], dtype=object)

In [5]:
df.select_dtypes(include="object").nunique()

id_pedido                    620000
id_cliente                   152176
fecha_pedido                    871
canal_compra                      4
metodo_pago                       4
ciudad_tienda                    12
categoria_producto                6
producto                         24
correo_cliente               518000
nivel_satisfaccion               15
nivel_lealtad                     4
comentario_cliente               12
fecha_actualizacion_stock      1177
dtype: int64

## Clasificación de mis columnas categóricas

**id_pedido**                   Nominal   620000 (Alta)

**id_cliente**                  Nominal   152175 (Alta)

**fecha_pedido**                Nominal   871 (Alta)

**canal_compra**                 Nominal   4 (Baja)

**metodo_pago**                  Nominal   4 (Baja)

**ciudad_tienda**               Nominal   12 (Baja)

**categoria_producto**           Nominal   6 (Baja)

**producto**                     Nominal   24 (Media)

**correo_cliente**               Nominal   518000 (Alta)

**nivel_satisfaccion**           Ordinal   15 (Media)

**nivel_lealtad**                Ordinal   4 (Baja)

**comentario_cliente**           Nominal   12 (Baja)

**fecha_actualizacion_stock**    Nominal   1177 (Alta)

## Codificadores 

### One-Hot Encoding

In [6]:
codificador_nominal = OneHotEncoder(sparse_output=False, handle_unknown="ignore")

In [7]:
codificador_nominal.fit(df[["categoria_producto", "metodo_pago", "canal_compra"]])

OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [8]:
#Añadimos dytpe para que salgan ceros y unos, de lo contrario solo va a salir True o False
pd.get_dummies(df, columns=["categoria_producto", "metodo_pago", "canal_compra"], prefix=["cat", "pago", "canal"], dtype=int)

,id_pedido,id_cliente,fecha_pedido,ciudad_tienda,codigo_postal,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,...,cat_Juguetería,cat_Moda,pago_Efectivo,pago_Paypal,pago_Tarjeta,pago_Transferencia,canal_App,canal_Marketplace,canal_Tienda,canal_Web
0,P695472,C150397,2025-06-04,Medellín,500103,Zapatos deportivos,644100.0,2,1288200.0,68,...,0,1,0,0,0,1,0,0,0,1
1,P290948,C127261,2025-07-17,Cali,7600103,Audífonos inalámbricos,74000.0,1,74000.0,53,...,0,0,0,0,1,0,0,0,0,1
2,P224930,C94844,2026-04-09,Bucaramanga,6800101,Set de cuchillos,956450.0,1,1809450.0,64,...,0,0,1,0,0,0,0,1,0,0
3,P456531,C125793,2025-05-31,Villavicencio,5000101,Audífonos inalámbricos,204200.0,0,0.0,40,...,0,0,0,0,1,0,0,1,0,0
4,P296569,C118158,2026-05-15,Ibagué,7300103,Perfume cítrico,493300.0,4,1809450.0,32,...,0,0,0,0,1,0,0,1,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
619995,P223511,C34096,2025-10-25,Cúcuta,5400101,Lámpara de escritorio,93900.0,2,187800.0,20,...,0,0,1,0,0,0,1,0,0,0
619996,P363057,C127294,2025-05-11,Cúcuta,5400102,Zapatos deportivos,155200.0,1,155200.0,31,...,0,1,0,0,1,0,0,1,0,0
619997,P593016,C32866,2026-12-01,Villavicencio,5000107,Set de bloques,221700.0,1,221700.0,61,...,1,0,0,1,0,0,0,0,1,0
619998,P398260,C37572,2025-08-27,Manizales,1700104,Pesas ajustables,800400.0,100,1809450.0,70,...,0,0,1,0,0,0,0,0,1,0


In [9]:
df.columns

Index(['id_pedido', 'id_cliente', 'fecha_pedido', 'canal_compra',
       'metodo_pago', 'ciudad_tienda', 'codigo_postal', 'categoria_producto',
       'producto', 'precio_unitario', 'unidades_vendidas', 'monto_compra',
       'edad_cliente', 'correo_cliente', 'nivel_satisfaccion', 'nivel_lealtad',
       'comentario_cliente', 'peso_pedido_kg', 'fecha_actualizacion_stock',
       'fuga_cliente', 'z_monto_compra', 'es_venta_corporativa',
       'es_precio_alto', 'edad_cliente_escalada', 'precio_unitario_escalada',
       'monto_compra_robusto', 'unidades_vendidas_robusto'],
      dtype='object')

### Ordinal Encoding

In [10]:
orden_lealtad = ["Bronce", "Plata", "Oro", "Platino"]

In [11]:
codificador_lealtad = OrdinalEncoder(categories=[orden_lealtad])

In [12]:
df["nivel_lealtad_cod"] = codificador_lealtad.fit_transform(df[["nivel_lealtad"]])

In [13]:
df[["nivel_lealtad", "nivel_lealtad_cod"]].drop_duplicates()

,nivel_lealtad,nivel_lealtad_cod
0,Oro,2.0
1,Platino,3.0
2,Plata,1.0
4,Bronce,0.0


## Por qué usé ordinal encoding en nivel_lealtad

Porque naturalmente esta variable tiene un orden en el contexto real y el de NovaMarket, entre más valioso es el nombre del mineral más valioso es el nivel de lealtad del cliente, si se asumen todos como iguales no tendría sentido llamarlo nivel

## Separación en entrenamiento y prueba

Antes de aplicar el target encoding, dividimos el dataset en dos partes: entrenamiento (80%) y prueba (20%). Esto se hace ahora porque el target encoding necesita calcularse solo con los datos de entrenamiento, si no habría fuga de datos.

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
df_entrenamiento, df_prueba = train_test_split(df, test_size=0.2, random_state=42)

`monto_compra` tiene algunos valores nulos que arrastramos de semanas anteriores. Los eliminamos solo del entrenamiento porque el `TargetEncoder` no acepta nulos en la variable objetivo.

In [16]:
df_entrenamiento = df_entrenamiento.dropna(subset=["monto_compra"])

In [17]:
print("Filas de entrenamiento:", len(df_entrenamiento))
print("Filas de prueba:", len(df_prueba))

Filas de entrenamiento: 479303
Filas de prueba: 124000


## Target Encoding de ciudad_tienda

Para `ciudad_tienda` usamos target encoding porque queremos que cada ciudad se represente según el monto de compra promedio de esa ciudad. Así, en vez de tener texto, cada ciudad queda como un número que refleja qué tan altos suelen ser sus montos de compra.

In [18]:
from category_encoders import TargetEncoder

In [19]:
codificador_ciudad = TargetEncoder(cols=["ciudad_tienda"], smoothing=10)
codificador_ciudad.fit(df_entrenamiento["ciudad_tienda"], df_entrenamiento["monto_compra"])

TargetEncoder(cols=['ciudad_tienda'])

El `smoothing=10` sirve para que las ciudades con pocos registros no queden con un valor muy raro; las suaviza acercándolas al promedio general.

### Transformar entrenamiento y prueba

Usamos el mismo codificador ya ajustado para transformar los dos conjuntos. No volvemos a hacer `fit()` sobre la prueba, solo `transform()`, así evitamos la fuga de datos.

In [20]:
df_entrenamiento["ciudad_tienda_cod"] = codificador_ciudad.transform(df_entrenamiento["ciudad_tienda"])
df_prueba["ciudad_tienda_cod"] = codificador_ciudad.transform(df_prueba["ciudad_tienda"])

In [21]:
print("Entrenamiento:")
print(df_entrenamiento["ciudad_tienda_cod"].describe())
print("Prueba:")
print(df_prueba["ciudad_tienda_cod"].describe())

Entrenamiento:
count    479303.000000
mean     624179.388508
std        2225.135707
min      620861.400622
25%      622188.074622
50%      623816.174118
75%      625306.444893
max      629134.715439
Name: ciudad_tienda_cod, dtype: float64
Prueba:
count    124000.000000
mean     624179.509301
std        2223.992478
min      620861.400622
25%      622188.074622
50%      623816.174118
75%      625306.444893
max      629134.715439
Name: ciudad_tienda_cod, dtype: float64


## Agrupamiento de codigo_postal

`codigo_postal` tiene alta cardinalidad (493 valores únicos), entonces no vale la pena aplicarle one-hot ni target encoding directamente. Lo que hacemos es agrupar los códigos poco frecuentes en una categoría llamada 'Otras', quedándonos solo con los códigos que aparecen más de 15 veces en entrenamiento.

In [30]:
conteo_postal = df_entrenamiento["codigo_postal"].value_counts()

In [31]:
postales_frecuentes = conteo_postal[conteo_postal >= 15].index

Aplicamos la misma lista `postales_frecuentes` (calculada solo con entrenamiento) también en la prueba, para no filtrar información.

In [32]:
df_entrenamiento["codigo_postal_agrupado"] = df_entrenamiento["codigo_postal"].where(df_entrenamiento["codigo_postal"].isin(postales_frecuentes), "Otras")
df_prueba["codigo_postal_agrupado"] = df_prueba["codigo_postal"].where(df_prueba["codigo_postal"].isin(postales_frecuentes), "Otras")

In [33]:
print("Códigos únicos originales:", df_entrenamiento["codigo_postal"].nunique())
print("Códigos únicos agrupados:", df_entrenamiento["codigo_postal_agrupado"].nunique())

Códigos únicos originales: 493
Códigos únicos agrupados: 303


## Exportación del dataset codificado

Guardamos por separado el conjunto de entrenamiento y el de prueba, para poder retomarlos en la siguiente semana sin volver a correr todo el notebook.

In [34]:
df_entrenamiento.to_csv("S09_PD_Camacho_Serpa_DatasetCodificado_entrenamiento.csv", index=False)
df_prueba.to_csv("S09_PD_Camacho_Serpa_DatasetCodificado_prueba.csv", index=False)

In [35]:
df_completo = pd.concat([df_entrenamiento, df_prueba])
df_completo.to_csv("S09_PD_Camacho_Serpa_DatasetCodificado.csv", index=False)